core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -564,6 +678,115 @@ public Map<K, V> read(Decoder decoder, Object reuse) throws IOException {
     }
   }
 
+  public abstract static class PlannedStructReader<S>
+      implements ValueReader<S>, SupportsRowPosition {
```

Comment:
This is the new base class that implements a read plan, which is a list of positions and readers. The read plan is produced by the new logic in `GenericAvroReader` that handles structs.